### PR TITLE
Limit description character length sent for health report

### PR DESCRIPTION
**Pull request by trstringer**
## Description

This PR limits the description length that can be sent for a health report. The platform supports a maximum of 4096 characters for the description text.

E.g. `<Description>this text can only be a max of 4096 characters</Description>`

Issue #2017 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/common/protocol/wire.py [hunk 1]
```diff
@@ -220,6 +220,12 @@ def _build_health_report(incarnation, container_id, role_instance_id, # pylint:
                          status, substatus, description):
     # Escape '&', '<' and '>'
     description = saxutils.escape(ustr(description))
+
+    # The max description that can be sent to WireServer is 4096 bytes.
+    # Exceeding this max can result in a failure to report health.
+    max_description_length = 4096
+    description = description[-max_description_length:]
```
Comment (johnsonshi) [inline azurelinuxagent/common/protocol/wire.py:227]: @trstringer Please log a warning if the description field is being trimmed. Also please leave a small but generous character buffer, as the platform usually mutates the values along the way (escaping and transforming chars). Thus the limit should be much lower than 4096.

Comment (trstringer) [reply azurelinuxagent/common/protocol/wire.py:227]: Once walinuxagent sends out the health status, why would the platform mutate any values? We already escape characters on the walinuxagent side, not sure I understand why that would happen again on the server side?

File: azurelinuxagent/common/protocol/wire.py [hunk 2]
```diff
@@ -220,6 +220,19 @@ def _build_health_report(incarnation, container_id, role_instance_id, # pylint:
...
+    # Exceeding this max can result in a failure to report health.
+    max_description_length = 4096
+    current_description_length = len(description)
+    if current_description_length > max_description_length:
+        logger.info(
+            'Trimmed health report description by {0} characters'.format(
+                current_description_length - max_description_length
+            )
+        )
+        description = description[-max_description_length:]
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/wire.py:234]: the truncation is being done after encoding the xml string... this has the potential of producing invalid xml if the encoding for a charater is truncated

Comment (trstringer) [reply azurelinuxagent/common/protocol/wire.py]: Good catch, you are correct. I've moved this trimming logic to before encoding to prevent this problem.

**Comment by narrieta** [comment]
code looks good; running automation

**Comment by trstringer** [comment]
Thanks, @narrieta!

**Review by narrieta — APPROVED** [review]

**Review by larohra — APPROVED** [review]

**Pull request merged**